### PR TITLE
Fixed nullability of enums in switches

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Fixes: https://github.com/dotnet/roslyn/issues/61594

### CODE REVIEW

**Comment by DoctorKrolic** [comment]
@CyrusNajmabadi Maybe review this as well?)

File: src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs [hunk 1]
```diff
@@ -1367,5 +1368,41 @@ void M()
...
+@"#nullable enable
+
+static class MyEnumExtensions
+{
+    public static string ToAnotherEnum(this MyEnum? myEnumValue)
+        => myEnumValue switch
+        {
+            MyEnum.Value1 => throw new System.NotImplementedException(),
+            MyEnum.Value2 => throw new System.NotImplementedException(),
+        };
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:1399]: Why null pattern isn't produced?

Comment (DoctorKrolic) [reply src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:1399]: Don't understand what exactly you mean. Can you give a piece of code?

Comment (Youssef1313) [reply src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:1399]: I was expecting the fixed code to be:

```csharp
#nullable enable
static class MyEnumExtensions
{
    public static string ToAnotherEnum(this MyEnum? myEnumValue)
        => myEnumValue switch
        {
            MyEnum.Value1 => throw new System.NotImplementedException(),
            MyEnum.Value2 => throw new System.NotImplementedException(),
            null => throw new System.NotImplementedException(),
        };
}

enum MyEnum
{
    Value1, Value2
}"
```

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:1399]: i'm fine either way.  up to you @DoctorKrolic if you want to enhance in that way.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs (hunk 1 above)
Comment (DoctorKrolic) [reply src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:1399]: The original issue is about wrong `?.` used, but I think it is preferable to implement Youssef1313's suggestion here as well

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:1399]: @DoctorKrolic Cool.   I'll wait on you for that.  Feel free to ping me when ready.

**Comment by DoctorKrolic** [comment]
@CyrusNajmabadi Done. That was a bit harder than I thought...

File: src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 2]
```diff
@@ -65,7 +76,7 @@ private static void HandleBinaryPattern(IBinaryPatternOperation? binaryPattern,
 
         private static void RemoveIfConstantPatternHasValue(IOperation operation, Dictionary<long, ISymbol> enumMembers)
         {
-            if (operation is IConstantPatternOperation { Value.ConstantValue: { HasValue: true, Value: var value } })
+            if (operation is IConstantPatternOperation { Value.ConstantValue: { HasValue: true, Value: var value } } && value is not null)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs:79]: nit:
```suggestion
            if (operation is IConstantPatternOperation { Value.ConstantValue: { HasValue: true, Value: not null and var value } })
```

up to you :)

File: src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchStatementHelpers.cs [hunk 3]
```diff
@@ -66,6 +66,25 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchOperation switch
...
+        public static bool HasNullSwitchArm(ISwitchOperation operation)
+        {
+            foreach (var switchCase in operation.Cases)
+            {
+                foreach (var clause in switchCase.Clauses)
+                {
+                    if (clause.CaseKind != CaseKind.SingleValue)
+                        continue;
+
+                    var value = ((ISingleValueCaseClauseOperation)clause).Value;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchStatementHelpers.cs:78]:
```suggestion
                    if (clause is not ISingleValueCaseClauseOperation { Value: var value })
                        continue;
```

File: src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchStatementHelpers.cs [hunk 4]
```diff
@@ -88,13 +107,19 @@ private static bool TryRemoveExistingEnumMembers(ISwitchOperation switchStatemen
...
-                            if (value == null || !value.ConstantValue.HasValue)
+                            if (value is null || !value.ConstantValue.HasValue)
                             {
                                 // We had a case which didn't resolve properly.  
                                 // Assume the switch is complete.
                                 return false;
                             }
 
+                            // null will be casted to 0, which creates a bug,
+                            // when a switch with null arm will not add enum's 0 value equivalentю
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchStatementHelpers.cs:118]:
```suggestion
                            // when a switch with null arm will not add enum's 0 value equivalent
```

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs (hunk 2 above)
Comment (DoctorKrolic) [reply src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs:79]: Uh, this fancy patterns)...

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**